open_lm/attention.py

```diff
@@ -33,6 +36,33 @@ def get_rectangular_mask(shape, q_seq_len, k_seq_len, device, dtype):
     return output_mask[:, :, :, :k_seq_len]
 
 
+def apply_attention_mask_(bias, attention_mask, queries_dtype):
+    """Applies attention mask (e.g., from HuggingFace generate) to an attention bias mask in-place.
+
+    Args:
+        bias (torch.Tensor, shape (batch_size, num_heads, q_seq_len, k_seq_len))
+        attention_mask (torch.Tensor, shape (batch_size, sequence_len))
+        queries_dtype: queries.dtype; used to get minimum value for masked indices.
+
+    Returns:
+        bias_with_mask (torch.Tensor, shape (batch_size, num_heads, q_seq_len, k_seq_len))
+    """
+    # Update mask to remove attention based on attention_mask that's passed in.
+    assert attention_mask.dim() == 2
```

Comment:
This forces the attention mask to be the same across all samples - while I understand the reasoning with this limitation (xformers is kind of weird about this I think), it might be worth adding an error message.